compiler/riscv/codegen/BinaryEvaluator.cpp

```diff
@@ -360,11 +360,79 @@ OMR::RV::TreeEvaluator::ishrEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    return shiftHelper(node, TR::InstOpCode::_sraw, TR::InstOpCode::_sraiw, cg);
    }
 
+// also handles bushr, sushr and lushr
 TR::Register *
 OMR::RV::TreeEvaluator::iushrEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return shiftHelper(node, TR::InstOpCode::_srlw, TR::InstOpCode::_srliw, cg);
-   }
+    TR::Node *firstChild = node->getFirstChild();
+    TR::Node *secondChild = node->getSecondChild();
+
+    TR::InstOpCode::Mnemonic opr;
+    TR::InstOpCode::Mnemonic opi;
+
+    if (node->getDataType().isInt64())
+       {
+       opr = TR::InstOpCode::_srl;
+       opi = TR::InstOpCode::_srli;
+       }
+    else
+       {
+       opr = TR::InstOpCode::_srlw;
+       opi = TR::InstOpCode::_srliw;
+       }
+
+    TR::Register *src1Reg;
+    TR::Register *trgReg;
+
+    if (cg->canClobberNodesRegister(firstChild))
+       {
+       src1Reg = trgReg = cg->gprClobberEvaluate(firstChild);
+       }
+    else
+       {
+       src1Reg = cg->evaluate(firstChild);
+       trgReg = cg->allocateRegister(TR_GPR);
+       }
+
+    /*
+     * If bitwidth is 8 or 16, we have to mask out high 48 or 56 bits
+     */
+    if (node->getDataType().isInt8())
+       {
+       generateITYPE(TR::InstOpCode::_andi, node, trgReg, src1Reg, 0xFF, cg);
+       src1Reg = trgReg;
+       }
+    else if (node->getDataType().isInt16())
+       {
+       generateITYPE(TR::InstOpCode::_slliw, node, trgReg, src1Reg, 16, cg);
+       generateITYPE(TR::InstOpCode::_srliw, node, trgReg, trgReg,  16, cg);
+       src1Reg = trgReg;
+       }
+
+    if (secondChild->getOpCode().isLoadConst() && secondChild->getRegister() == NULL)
+       {
+       int64_t value = secondChild->getLongInt();
+       if (VALID_ITYPE_IMM(value))
+          {
+          generateITYPE(opi, node, trgReg, src1Reg, value, cg);
+          }
+       else
+          {
+          TR::Register *src2Reg = cg->evaluate(secondChild);
+          generateRTYPE(opr, node, trgReg, src1Reg, src2Reg, cg);
+          }
```

Comment:
Only the bottom 5/6 bits of the shift amount are actually used, so it should be safe to just mask out the upper bits and always use the I-type instruction.

Also, is using `VALID_ITYPE_IMM` even valid here in the first place? It looks to me like the RISC-V ISA says that the upper bits of the `imm` field aren't supposed to be used to encode a shift amount and doing so could result in an invalid instruction.

```suggestion
       value &= node->getDataType().isInt64() ? 0x3f : 0x1f;
       generateITYPE(opi, node, trgReg, src1Reg, value, cg);
```

Reply:
You're right. Arithmetic shifts suffer the same problem so we have to fix `shiftHelper()` in similar manner. Moreover, truncation can be done with fewer instructions when shift amount is statically known. Will fix that. Thanks!